### PR TITLE
Exclusion improvements

### PR DESCRIPTION
**Pull request by Lalem001**
An attempt at improving file exclusion.

The biggest improvement appears when using the `--all` flag.

Changes:
- Always exclude `node_modules`
- Exclude using glob/minimatch patterns instead of RegExp (breaking change)
- addAllFiles now excludes during glob.sync, significantly improving performance

Resolves:
- #56: Exclude now uses glob patterns via minimatch

Complicates:
- #50: node_modules are now always excluded

Feel free to question my changes/decisions. I am open to compromise if necessary.

### CODE REVIEW

**Comment by bcoe** [comment]
@Lalem001 I dig these changes :+1: a couple asks:

* would it be possible to add one unit test that demonstrates the glob-based exclusions in action., perhaps mock out the compile step and ensure that the appropriate set of files gets flagged.
* could we update the README with an example of the glob based approach, perhaps a common config somewhat might use?

Once we get this merged, and add the includes functionality let's publish this as 5.x.

**Comment by Lalem001** [comment]
You ok with more commits? Or prefer squashed single commit?

**Comment by bcoe** [comment]
@Lalem001 extra commits are fine, I tend to only squash in extreme cases, e.g., I've made 99 silly mistakes before opening a pull, in the general case I prefer to have a record of what actually happened.

**Comment by Lalem001** [comment]
Regarding the communications toward issue #50, I was thinking maybe we can implement something similar to ESLint with exclusion negations? Thoughts?

**Comment by bcoe** [comment]
@sindresorhus does this branch address the issues you raised in:

https://github.com/bcoe/nyc/issues/56

and

https://github.com/bcoe/nyc/issues/57

Mind giving it a spin?

**Comment by bcoe** [comment]
@Lalem001 hey, just remembered we already use `glob` in the codebase, I don't think we need `glob` and `minimatch` ... perhaps you could replace our usage of `glob` with `minimatch` instead?

**Comment by Lalem001** [comment]
glob uses minimatch internally. I can look to see if I can access minimatch via glob in the morning.

**Comment by bcoe** [comment]
@Lalem001 gotcha, sorry for being confusing, I was thinking that `minimatch` was a stripped down version of `glob` (like optimist vs. minimist); I'd rather pull in both dependencies than jump through hoops to make `glob` work as our string matcher.

**Comment by bcoe** [comment]
@Lalem001 let's give @sindresorhus a chance to chime in on this, since it solves a couple issues he opened, and then we'll get this merged :+1:

File: README.md [hunk 1]
```diff
@@ -104,20 +104,32 @@ nyc report --reporter=lcov
...
+any files with the extension `.spec.js`, and anything in the `build/`
+directory:
+
+```json
 {"config": {
   "nyc": {
     "exclude": [
-      "node_modules/"
+      "**/*.spec.js",
+      "build/**"
```
Comment (sindresorhus) [inline README.md]: According to line 75 the `/**` is not needed

File: index.js [hunk 2]
```diff
@@ -65,6 +64,24 @@ NYC.prototype._createInstrumenter = function () {
...
+NYC.prototype._prepExcludePatterns = function (excludes) {
+  var directories = []
+  excludes = _.map(excludes, function (exclude) {
+    // Remove leading "current folder" prefix
+    if (_.startsWith(exclude, './')) {
+      exclude = exclude.slice(2)
+    }
+
+    // Allow gitignore style of directory exclusion
+    if (!_.endsWith(exclude, '/**')) {
```
Comment (sindresorhus) [inline index.js]: What if it ends with `/` like the `node_modules/` in the readme example? Then it will become `node_modules//**`.

File: index.js [hunk 3]
```diff
@@ -30,11 +31,9 @@ function NYC (opts) {
   config = config.nyc || {}
 
   // load exclude stanza from config.
-  this.exclude = config.exclude || ['node_modules[\/\\\\]', 'test[\/\\\\]', 'test\\.js']
+  this.exclude = ['node_modules/**'].concat(config.exclude || ['test/**', 'test.js'])
```
Comment (sindresorhus) [inline index.js]: I would ignore `test-*.js` by default too. It's a very common pattern. https://github.com/sindresorhus/ava/blob/eaf3e0f8a66980e87496a30b9f4a8c45329f0ae9/api.js#L198

File: index.js [hunk 4]
```diff
@@ -65,6 +64,24 @@ NYC.prototype._createInstrumenter = function () {
   })
 }
 
+NYC.prototype._prepExcludePatterns = function (excludes) {
+  var directories = []
+  excludes = _.map(excludes, function (exclude) {
+    // Remove leading "current folder" prefix
```
Comment (sindresorhus) [inline index.js]: Why is this needed? Pretty sure `glob` handles this fine.

File: index.js (hunk 3 above)
Comment (sindresorhus) [inline index.js]: What if you also have `node_modules` directories in sub-directories? Shouldn't the pattern be `**/node_modules/**`?

https://github.com/sindresorhus/ava/blob/eaf3e0f8a66980e87496a30b9f4a8c45329f0ae9/api.js#L203

File: README.md [hunk 5]
```diff
@@ -104,20 +104,32 @@ nyc report --reporter=lcov
...
+the array is a glob pattern indicating which paths should be omitted.
 
-```js
+Globs are matched using [minimatch](https://github.com/isaacs/minimatch)
+
+In addition to patterns specified in the package, nyc will always ignore
+files in `node_modules`.
+
+For example, the following config will ignore all of `node_modules`,
+any files with the extension `.spec.js`, and anything in the `build/`
```
Comment (sindresorhus) [inline README.md]: `build/` => `build`, for consistency

File: index.js (hunk 3 above)
Comment (bcoe) [inline index.js]: Agreed, it seems reasonable to assume we'd never want to instrument in `node_modules` regardless of the nesting.

**Comment by bcoe** [comment]
@Lalem001 probably worth adding a few tests for the scenarios @sindresorhus brought up:

`node_modules/`,  `/foo/node_modules/foo`, and `./foo`.

In the test for `./foo`, you can check whether or not removing the leading `./` is required :+1:

**Comment by bcoe** [comment]
@Lalem001 I should add, sorry for all the back and forth! your work is looking great.

File: index.js (hunk 2 above)
Comment (Lalem001) [inline index.js]: This was done to mimic behavior seen in ESLint's ignore methods. This may actually be an issue with ESLint as well.

Would you guys prefer I went with strict glob? No pattern adjustments at all?

File: index.js (hunk 4 above)
Comment (Lalem001) [inline index.js]: Again, something I saw in ESLint's source.

**Comment by es128** [comment]
Just to chime in with a quick suggestion, the breaking change with dropping support RegExp could be avoided by using an interface that supports both globs and RegExp such as [anymatch](https://github.com/es128/anymatch).

Using a [bound matcher function](https://github.com/es128/anymatch#anymatch-matchers) in place of `shouldInstrumentFile` will also improve performance, preventing the need for minimatch to re-parse each glob on each file.

**Comment by Lalem001** [comment]
@es128 `anymatch` would work as a `minimatch` replacement, but `glob` still runs on `minimatch`.

The primary improvement coming from this PR comes from the way I pass the patterns to `glob` when it is getting all files in the directory. As far as I know glob will not accept REGEX patterns.

File: index.js (hunk 4 above)
Comment (Lalem001) [inline index.js]: Just did some quick testing with minimatch itself

```javascript
// minimatch(path, pattern)

minimatch('./node_modules/foo.js', 'node_modules/**') // false
minimatch('node_modules/foo.js', './node_modules/**') // false
minimatch('node_modules/foo.js', 'node_modules/**')   // true
```

Perhaps I should do as ESLint and also remove `./` from the path that is passed in?

**Comment by es128** [comment]
Good point. The maintainers here don't seem concerned about the breaking change, so it's probably best left with the approach you have at this point.

But if there is any motivation to preserve support for regex, one more suggestion would be to also use [readdirp](https://github.com/thlorenz/readdirp) instead of node-glob and leverage its ability to handle functions for filtering. There may be a bit of perf boost to that change as well.

2¢

**Comment by bcoe** [comment]
@es128, @Lalem001, I feel that globs are probably the more expected behavior for supporting exclusions -- I used regexes as a naive first pass at the problem.

I think we should land this update on `nyc@5.x` and if allow future pull requests to ask for regex support :+1: -- as is, switching to glob-based matching solves a few open issues.

@Lalem001 if you don't mind addressing Sindre's comments (adding a couple extra tests, I'm in a agreement that we should also ignore test-*.js), I'll get this merged tonight.

File: index.js (hunk 2 above)
Comment (sindresorhus) [inline index.js]: I didn't comment on the merit of it. Just that I would handle a trailing `/` better by first removing it.

Comment (Lalem001) [inline index.js]: Ok, I will remove any trailing `/` before appending `/**`

**Comment by Lalem001** [comment]
I would like to point out a circular dependency between `nyc` and `tap`.

Tap depends on `nyc@3.1.0`. Thus the root package still should to specify `node_modules` until tap is updated.

**Comment by Lalem001** [comment]
This is ready for another review

**Comment by bcoe** [comment]
@Lalem001 :+1: I will make an effort to get things merged tonight.

**Comment by bcoe** [comment]
@Lalem001 I haven't had too much of a chance to debug yet, but I'm noticing that running coverage takes  a very long time on some libraries now, I tested on the following:

https://github.com/sindresorhus/pageres
https://github.com/kriskowal/q
https://github.com/bcoe/yarsay

The problem seems to relate to the default exclusion rules not properly disregarding `node_modules`:

screen shot 2015-12-02 at 8 07 18 pm

**Comment by Lalem001** [comment]
I am looking into it now. Thanks for the heads up.

**Comment by Lalem001** [comment]
Did some preliminary debugging.

I can reproduce the issue you observed when I `npm link nyc` into `yarsay`. In this case, the `node_modules` getting coverage are the `node_modules` of nyc.

When I installed nyc via `npm install <githubname>/<githubrepo>[#<commit-ish>]`, everything appeared to work as expected.

Did you install via another method?

Should we attempt to handle this case?

**Comment by Lalem001** [comment]
Perhaps this is #37 reoccurring?

**Comment by Lalem001** [comment]
Ultimately the issue is the fact that the offending relative paths lead with `../` and that `**/node_modules/**` will not match such paths.

One solution would be to remove leading `../` from all relative paths.
`Path.normalize` might handle this for us.

**Comment by bcoe** [comment]
@Lalem001 I think this would be worthwhile. Having the `npm link` flow work is great for testing, so I think we should make sure to support this.

**Comment by Lalem001** [comment]
@bcoe I apologize in advance for the wall of text.

I have experimented with the proposed solutions. I have had some success with sending both the relative and full path to `shouldInstrumentFile`. If either path is excluded, then I exclude the file.

There is a caveat however. If, by chance a js file in the root of the symlinked module was required, then that file would not be excluded by the `**/node_modules/**` rule.

For example, if a project under test, symlinked glob, the paths could be something like:

- `../glob/glob.js` -> included
- `/Users/user/glob/glob.js` -> included

As you can see, `node_modules` does not appear in these paths.

In the mean time, glob's references to modules would be excluded.

- `../glob/node_modules/minimatch/minimatch.js` -> included
- `/Users/user/glob/node_modules/minimatch/minimatch.js` -> **excluded**

If you are ok with this caveat, I am ready to commit/push.

---

If not, I have an idea that would transform symlink target paths to paths that are more "logical", but it would likely require a small performance hit.

Basically I would glob the full project searching for all symlinks. Then compare each file's path to a list of known symlink targets. If there is a match, replace the matching portion of the file's path with the symlink's path.

example:
Given a symlink: `project/node_modules/nyc -> /Users/user/nyc`
```javascript
var path = require.resolve('nyc')
// > '/Users/user/nyc/index.js'
path = path.replace('/Users/user/nyc', 'project/node_modules/nyc')
// > 'project/node_modules/nyc/index.js'
```

The full project glob `**/*/`, can likely be reduced to `node_module/*/` glob, significantly reducing the performance hit, but may let edge cases through.

**Comment by Lalem001** [comment]
A somewhat related, open, issue can be found in https://github.com/nodejs/node/issues/3402

**Comment by bcoe** [comment]
@Lalem001 thanks for all the hard work. This caveat sounds reasonable, and should considerably increase the speed of running `nyc` when link has been used. If you push up code today, I'm hoping to do some OSS work today.

**Comment by Lalem001** [comment]
I am not sure why these tests are failing. Even my local copy of latest master is failing. Ideas?

**Comment by bcoe** [comment]
@Lalem001 merging two branches broke each other in conjunction, I'll have a patch in shortly.

**Comment by Lalem001** [comment]
@bcoe Travis tests are passing now

**Comment by bcoe** [comment]
Closing in favor of https://github.com/bcoe/nyc/pull/84, which has a few minor tweaks.
